### ISSUE TITLE
Function singleSearchStep was renamed to _singleSearchStep not everywhere

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4178]
|<img alt="natta's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| natta reported|
|-|-|
|Trac ID|trac#4178|
|Type|release blocker: regression|
|Created|2009-12-19 16:13:42Z|
|Branch|https://github.com/twisted/twisted/tree/imap-search-fixes-4178|

Function singleSearchStep was renamed to _singleSearchStep and one parameter added, but in search_OR and search_NOT fucntions there is still old version function invocation (i.e. without "_")
So every search query that contains OR or NOT keywords raises errors.

Attachments:

* [imap4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e90/e90a987a42c0262c4bf70466d5bc7272ec86844f/c1ae833211cf2601568f81218512337ff1c4ae59.patch) (454 bytes) - added by natta on 2009-12-19 16:14:59Z - Patch that replaces singleSearchStep with _singleSearchStep in search_(NOT|OR)

<details><summary>Searchable metadata</summary>

```
trac-id__4178 4178
type__release_blocker__regression release blocker: regression
reporter__natta natta
priority__high high
milestone__Twisted_10_0 Twisted-10.0
branch__branches_imap_search_fixes_4178 branches/imap-search-fixes-4178
branch_author__exarkun__itamar exarkun, itamar
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__mail_imap4 mail imap4
time__1261239222000000 1261239222000000
changetime__1266695509000000 1266695509000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="natta's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|natta set owner to natta<br>natta set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="natta's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|natta set owner to @exarkun<br>natta set status to `new`|
|-|-|

I can not commit changes, so i reassign it back

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Regression introduced by [#1977](https://github.com/twisted/twisted/issues/1977).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to natta|
|-|-|

Thanks for the patch, natta.  Can you supply one which also includes unit tests for this fix?  `twisted.mail.test.test_imap.DefaultSearchTestCase` may be a good case to extend.

**Comment by twisted-trac** [comment]
|<img alt="natta's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|natta commented|
|-|-|

Replying to [exarkun](#note_4):
> Thanks for the patch, natta.  Can you supply one which also includes unit tests for this fix?  `twisted.mail.test.test_imap.DefaultSearchTestCase` may be a good case to extend.

OK. As so I am not a very experienced developer, but I'll try.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [28242]) Branching to 'imap-search-fixes-4178'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner|
|-|-|

Fixed and tested, please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @exarkun|
|-|-|

So I would prefer another way to check if the lastSequenceId parameter is required. Something like that:
```
_requiresLastSequenceId = set(["NOT", "OR"])

if c in self._requiresLastSequenceId:
     result = f(query, id, msg, lastSequenceId)
else:
     result = f(query, id, msg)

```

The tests need to fit in 80 colums, too.

Please merge after that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [28281]) Merge imap-search-fixes-4178

Author: exarkun, itamar
Reviewer: therve
Fixes: [#4178](https://github.com/twisted/twisted/issues/4178)

Fix the regression in search OR and NOT support in the IMAP4 server's fallback
search support.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
